audio/internal/convert/stereo_test.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2024 The Ebitengine Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package convert
+
+import (
+	"bytes"
+	"io"
+	"math"
+	"reflect"
+	"testing"
+)
+
+type testCase struct {
+	resolution int
+	mono       bool
+	input      []byte
+	expected   []byte
+}
+
+func freq(n int) float64 {
+	return float64(n) * math.Pi / 2
+}
+
+func writeWord(w *bytes.Buffer, bits int, v uint32) {
+	v = v >> (32 - bits)
+	for i := 0; i < bits/8; i++ {
+		b := byte(v & 0xFF)
+		w.WriteByte(b)
+		v = v >> 8
+	}
+}
+
+func genSin(sinLen int, bits int, mono bool) []byte {
+	w := bytes.NewBuffer(nil)
+	for i := 0; i < sinLen; i++ {
+		v := uint32(float64((1<<31)-1) * (1 + math.Sin(freq(i))))
+		writeWord(w, bits, v)
+		if !mono {
+			writeWord(w, bits, v)
+		}
+	}
+	return w.Bytes()
+}
+
+func genTestCases() []testCase {
+	res16 := genSin(4, 16, false)
+	res8 := []byte{}
+	for i, v := range res16 {
+		if i%2 == 0 {
+			v = 0
+		}
+		res8 = append(res8, v)
+
+	}
+
+	cases := []testCase{
+		{8, true, nil, res8},
+		{8, false, nil, res8},
+		{16, true, nil, res16},
+		{16, false, nil, res16},
+		{24, true, nil, res16},
+		{24, false, nil, res16},
```

Comment:
Why don't we have res24?

Reply:
AFAICT, the whole point of `Stereo` is that it converts 8/16[/24/32] bit samples into 16 bit stereo samples. I was just keeping the same functionality as previous. If that understanding is correct, then 16bit values are the gold standard. Now that doesn't work for 8bit samples, as we just don't have the low-bits of resolution, but I'm just trying to create the same functionality.

Incidentally, 2 small notes:

- 16 bits of resolution is 96db of SNR (at full scale). That should be plenty, and using 16bits throughout the sound system is probably adequate.

- The original conversion code looks incorrect to me. For example, the [8bit code](https://github.com/hajimehoshi/ebiten/blob/main/audio/internal/convert/stereo16.go#L56) duplicated the byte high and low (constant `0x101`). While that's a solution, I think it's less surprising than just calling the low-bits 0.

Reply:
> AFAICT, the whole point of Stereo is that it converts 8/16[/24/32] bit samples into 16 bit stereo samples.

Yes that's correct. Stereo16 (the original name) indicates that the stereo 16bit (little endian) PCM source, whatever the original source is. So, 8bit code is multiplied by 0x101 to convert 8bit to 16bit values.

Please revert your renaming (Stereo16 to Stereo).

> Now that doesn't work for 8bit samples, as we just don't have the low-bits of resolution, but I'm just trying to create the same functionality.

Do you mean your current test res8 doesn't work, or the current implementation (before your PR) doesn't work?

If we think the current implementation doesn't work with 8bit wav, we should focus on this problem first.

Reply:
I've confiemd 8bits worked correctly with the current implementation. https://github.com/hajimehoshi/ebiten/commit/1cd6a1f10ce6ec5251bc0834f135ded0b92c4269 It's because samples in 8bit wav file are unsigned, not signed.

Then, please revert your implementation for 8bits. Thanks,